feat: Add plugin to create badges.

WinnerWind opened the pull request:
This is a PR for the inclusion of a new plugin, which may be useful for users. This is a port of the [Obsidian badges plugin](https://github.com/gapmiss/badges)

# Badges
Create beautiful badges with simple syntax. Useful for making text pop out, and to include the occasional thingamagig.
![image](https://github.com/user-attachments/assets/ed43aacf-5a61-4a58-a5aa-3a8ad71ad87b)
![image](https://github.com/user-attachments/assets/0f224860-e09f-4802-9349-f9ea311d4384)

[See it in action](https://winnerwind.in/Blogs/Tutorials/NAT-Bypass#7-subdomains)
## About
This plugin is a transformer plugin that allows the user to make "badges" with relative ease. To make a badge, an inline code block is used. Code blocks are detected using REGEX. Despite being a port of an Obsidian plugin, it is not dependent on `ofm`[^1]. I have done my best to keep it as faithful to the original, however any changes to the original will require this to be changed manually.
[^1]: Unfortunately, for proper display, an [obsidian theme](https://github.com/saberzero1/quartz-themes/tree/master?tab=readme-ov-file) is required. I'm working on lifting this limitation.

## Package Dependencies
- Lucide Static (`lucide-static`) (Already included in `package.json`)
## Syntax
```
`[!!KEY:TEXT]` -> Normal badges, see "About"
`[!!|GHS>SUBTEXT:TEXT]`  -> Github Success
`[!!|GHB>SUBTEXT:TEXT]` -> Github Blue
`[!!|KEY:VAL]` -> Plaintext
`[!!|ICON|ARIA:TEXT|COLOR-RGB]` -> Custom icon. Color can be a CSS Variable, eg. `var(--color-red-rgb)`
```

WinnerWind committed:
Add badges plugin with code and CSS.

saberzero1 commented on content/.gitkeep:
Don't delete this.

WinnerWind replied:
Oh, not sure why that got deleted. Will fix.

WinnerWind replied:
Added it back.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+import { QuartzTransformerPlugin } from "../types"
+import badgesCSS from "../../styles/badges.scss"
+// import badgesFixCSS from "../../styles/badges.fix.css"
```
Remove commented code

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+// import badgesFixCSS from "../../styles/badges.fix.css"
+import icons from "lucide-static";
+
+export const BADGE_TYPES:any[] = [
```
Is there a particular reason for the chosen icons? (Beside the ones we use for callouts)

WinnerWind replied:
I copied this from the source plugin. That one uses these icons, so I copied them for consistency.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+	["vault", 'Vault', 'vault'],
+];
+
+const REGEXP = /\[!!([^\]]*)\]/gm;
```
`[!!]` will also match. Is this edge-case accounted for?

WinnerWind replied:
This edge case is not accounted for. It returns `Badges syntax error`, which I guess does make sense, but I'll try to rewrite the regex (i'm not very experienced).
Also, from my memory, the original plugin uses the exact same regex.

WinnerWind replied:
I just checked the original plugin, and it seems that this edge case is accounted for in the original but not my port. Will look into this.

WinnerWind replied:
Edge case has been accounted for with new regex.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+
+function buildBadge(text: string) {
+  // HTML Elements
+  let newEl:string = "";
+  let iconEl:string = "";
+  let titleEl:string = "";
+  let textEl:string = "";
```
`""` is implicitly `string`.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+  let iconEl:string = "";
+  let titleEl:string = "";
+  let textEl:string = "";
+  let attrType:any = "";
```
This is implicitly a string.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+  let titleEl:string = "";
+  let textEl:string = "";
+  let attrType:any = "";
+  let part:string = text.substring(2);
```
Can be `null`.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+  let textEl:string = "";
+  let attrType:any = "";
+  let part:string = text.substring(2);
+  let content:string = part.substring(part.length-1,1).trim();
```
Can be `null`.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+    return newEl
+  }
+
+  let parts:any[] = content.split(':');
```
Implictly list of strings.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+  let parts:any[] = content.split(':');
+  // return if NO CONTENT
+  if (parts.length < 2) {
+    newEl = "<span style=\"color:var(--text-error)\">❌ Badges syntax error</span>";
```
Use backticks to prevent needing to escape. E.g. `<span style="..."></span>`.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+  // type of badge
+  let badgeType:string = parts[0].trim();
+  // build and check for extras
+  let extras:any[] = badgeType.split("|");
```
Implicitly list of strings.

WinnerWind commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+const DOCUMENTCODE = 'code'
+
+// This plugin has no options.
+export interface Options {
```
Removing this unused code is impossible because then the plugin cannot compile.

WinnerWind commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+  let textEl:string = "";
+  let attrType:any = "";
+  let part:string = text.substring(2);
+  let content:string = part.substring(part.length-1,1).trim();
```
Not sure if this is helpful.

WinnerWind commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+  let badgeContent:string = parts[1].trim().split('|')[0] //Original code doesnt have .split[...] but for some reason its needed for proper rendering.
+
+  // custom badge
+  if (extras.length == 3) {
```
I don't understand how this is wrong. This is consistent with the original plugin.

WinnerWind commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,241 @@
+function getLucideIconSvg(iconName: string): string {
+	var potentialIconReturn = icons[toPascalCase(iconName.replace(/^lucide-/, '')) as keyof typeof icons];
+	if (!potentialIconReturn == ""){
+		var iconToReturn = icons[toPascalCase(iconName.replace(/^lucide-/, ''))].replace(/\n/g, "").replace("  ", " ")
```
I didn't even notice this. Great suggestion, fixed it.

WinnerWind commented:
Should be ready to merge. I'll be squashing the commits.

jackyzha0 commented:
@saberzero1 let me do a review pass before merging ty, will get to it later today

WinnerWind commented:
No rush. Good luck!

WinnerWind committed:
Allow definition of custom badges via settings

Custom badges can be defined via `quartz.config.ts` by adding the
`customBadge` option, an array. Eg.
```
      Plugin.InlineBadges({
        customBadges: [["menu", "menu-test", [140, 200, 110,1], 0.117],
                        ["toggle-right","switch-on",[225, 0,
40,1],0.117]
                        ],
        }),
```

Array is defined in the format
```
 [ [icon,name,[RED,GREEN,BLUE,ALPHA],TEXT_ALPHA],
[icon,name,[RED,GREEN,BLUE,ALPHA],TEXT_ALPHA] ]
```

WinnerWind committed:
Add type safety to customBadges

please be happy github actions

`integer` not `int`

How did my code compile?

Use proper touple syntax

Ok

Fix array type for customBadges

this took too long.

WinnerWind commented:
Hey all, just added an option `customBadges` to define custom badges for the syntax `[!!type:Text]`.

WinnerWind committed:
Add an option to include CSS for proper rendering

Now, you don't need an Obsidian Theme to use this plugin.

WinnerWind committed:
Fix code style

WinnerWind commented:
Added an option to include some Obsidian CSS if the user needs the badges to render properly but don't use an Obsidian theme.
Example line in `quartz.config.ts`
```ts
Plugin.InlineBadges({
  customBadges: [["menu", "menu-test", [140, 200, 110,1], 0.117],
                  ["toggle-right","switch-on",[225, 0, 40,1],0.117]
                  ],
  useObsidianCSS: true
  }),
```
I think that the current implementation is satisfactory. Thoughts?

I don't think I'll be adding any more features to this right now.

EDIT : Defaults
```ts
Plugin.InlineBadges({
  customBadges: []
  useObsidianCSS: false
  }),
```

saberzero1 commented:
> Added an option to include some Obsidian CSS if the user needs the badges to render properly but don't use an Obsidian theme.
> Example line in `quartz.config.ts`
> ```ts
> Plugin.InlineBadges({
>   customBadges: [["menu", "menu-test", [140, 200, 110,1], 0.117],
>                   ["toggle-right","switch-on",[225, 0, 40,1],0.117]
>                   ],
>   useObsidianCSS: true
>   }),
> ```
> I think that the current implementation is satisfactory. Thoughts?
> 
> I don't think I'll be adding any more features to this right now.
> 
> EDIT : Defaults
> ```ts
> Plugin.InlineBadges({
>   customBadges: []
>   useObsidianCSS: false
>   }),
> ```

I feel like you might want to use a list of Objects for configuration. E.g. something like this:

```ts
Plugin.InlineBadges({
  customBadges: [
    {
      name: "menu",
      icon: "menu-test",
      color: rgba(140, 200, 110, 1),
      opacity: 0.117
    },
    {
      name: "toggle-right",
      icon: "switch-on",
      color: rgba(255, 0, 40, 1),
      opacity: 0.117
    }
  ],
  useObsidianCSS: false
}),
```

WinnerWind commented:
I feel like using a dictionary may be superior. Will look into its implementation. The current implementation is personally very easy for me to modify and work with, and also ensures that *all* the data is filled properly.

A dictionary does sound superior, though it does come at the cost of the user accidentally leaving data. 

I'll give it a shot.

WinnerWind committed:
Use a dictionary instead of an array.

Should make it easier to define badges.

WinnerWind commented:
I'm done. It uses an object which is defined like so
```ts
  customBadges: Array<{
    name: string
    icon: string
    color: [number, number, number, number]
    textOpacity: number
  }>
```
Hope this is satisfactory!

WinnerWind commented:
Just thought I'd check in.
Are there any changes you'd like for me to add?

saberzero1 commented on quartz/styles/badges-optional.scss:
Why is this file named as optional, but is not optional?

WinnerWind replied:
It's optional because you don't NEED to include it if you use an Obsidian Theme for Quartz.

saberzero1 commented on quartz/styles/badges.scss:
```diff
@@ -0,0 +1,421 @@
+.inline-badge[data-inline-badge="info"],
+.inline-badge[data-inline-badge="todo"],
+.inline-badge[data-inline-badge="note"] {
+  color: rgba(var(--color-blue-rgb), 1);
+  background-color: rgba(var(--color-blue-rgb), 0.123);
```
Perhaps it would be better to match Quartz's callout colors instead of injecting Obsidian color variables?

WinnerWind replied:
I just copy pasted that CSS from the original plugin in order to maintain parity with the original, but you do have a point.

By using Quartz' callout colours, we can get rid of the need to include an Obsidian Theme.

I'll see what I can do about this.

WinnerWind replied:
Okay, so I've looked over everything again. The issue with this is:
1. I'd have to refactor how the entire plugin works, because currently the callout colour variables are only available to `&[data-callout]`. Which means some potential code refactoring.
2. I'd need to write the CSS from scratch, which isn't too much of a burden for me but it is certainly quite a task (there's nearly 422 lines here, and I'm a student so this part may take me a few weeks)

Though, if you don't mind waiting for a bit, I'm willing to get working on it. Also, this does mean that every single badge will be a `data-callout` object, so I am very worried about what this means for future plugins. 

Alternately, we could expose the colours in `:root` which would let me access it despite it being a `data-callout` object. 

Let me know if my understanding of how scss works is wrong, and if there's really a way to access the colour variables in `data-callout`.

saberzero1 replied:
I moreso meant using the Quartz callout colors.

So instead of the value Obsidian uses for `--color-orange`, we would use the orange that is used in some Quartz callouts. Same for all other colors Obsidian defines. (Red, orange, yellow, green, cyan, blue, purple, and pink)

WinnerWind replied:
Ah okay, so you want me to hard code the colours, but keep them in sync with Quartz' colours. I think I can do that. 

I think this would just mean to get rid of `badges-optional.scss`, which is meant to feed the Obsidian variables into the final CSS.

WinnerWind replied:
<img width="889" height="68" alt="image" src="https://github.com/user-attachments/assets/c1a8efad-9712-49c5-b085-03a7577b6d59" />
<img width="887" height="60" alt="image" src="https://github.com/user-attachments/assets/20c95503-5d4a-499c-81e0-1e422a91ac05" />


Set the colours and only included the absolute essential parts of `badges-optional.scss`. This look okay?

Note that this was rendered without `useObsidianCSS`.

saberzero1 commented on package.json:
```diff
@@ -110,6 +110,7 @@
     "tsx": "^4.19.4",
-    "typescript": "^5.8.3"
+    "typescript": "^5.8.3",
+    "lucide-static": "0.511.0"
```
Your PR description mentions that `lucide-static` could be used because it is already included, but it seems you added it yourself.

WinnerWind replied:
It's already included? I didn't see it in `package.json` before I added it myself.

WinnerWind replied:
Oh, it took me a bit to understand what you meant by that.

@saberzero1  what I meant to say in that line "already included" was that I was including the change to `package.json` within my PR, not that it was already part of `package.json`. sorry for the confusion. 

In short, currently `lucide-static` is NOT included. My PR *will* include `lucide-static`.

saberzero1 commented:
Also, this plugin is currently undocumentated.

WinnerWind commented:
How do you suggest I write documentation?

WinnerWind committed:
Remove the need for Obsidian Optional CSS

this CSS is now integrated with the normal css file, which uses quartz
colours by default.

WinnerWind committed:
Move SVG alignment code to `badges.scss`

WinnerWind committed:
Add documentation

WinnerWind commented:
I've written documentation under `docs/plugins/InlineBadges.md`, and I've applied the requested changes. Please let me know whether they are to your satisfaction @saberzero1 .

WinnerWind commented:
Just noticed that the InlineBadges are rendering within code blocks. Will need to fix that.

WinnerWind committed:
Block multi-line code blocks from being caught

This prevents multi-line code blocks containing a badge definition to be
replaced with badge HTML which is certainly unintended.

WinnerWind commented:
Okay, this should work. Let me know if there are any changes you want from your end. 

EDIT: would be really embarrassing if in the end this code didn't work

saberzero1 commented on docs/plugins/InlineBadges.md:
````diff
@@ -0,0 +1,79 @@
+
+A badge object is defined like so:
+
+```ts
+{
+	name: "foobar",
+	icon: "icon_name",
+	color: [10, 10, 10, 1],
+	textOpacity: 0.117,
+}
+```
````
````suggestion
```ts
{
  name: "foobar",
  icon: "icon_name",
  color: [10, 10, 10, 1],
  textOpacity: 0.117,
}
```
````

Consistent spacing with other docs.

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,281 @@
+import { JSResource, CSSResource } from "../../util/resources"
+import icons from "lucide-static"
+
+export const BADGE_TYPES: any[] = [
```
Could have a proper type, considering the structure is very clear. (array of arrays of ID, name, icon)

WinnerWind replied:
Set the type to `[string, string, string][]`

saberzero1 commented on quartz/plugins/transformers/badges.ts:
```diff
@@ -0,0 +1,281 @@
+  ["vault", "Vault", "vault"],
+]
+
+var allBadges: any[] = BADGE_TYPES //Append custom badges to the end of this array.
```
Same here

WinnerWind replied:
Set the type to `[string, string, string][]`

saberzero1 reviewed (commented):
Small remark on the docs.

As a final thought: in the badges configuration, perhaps another, optional parameter could be added to specify/override the color of the badge of that type? That way, user-defined badges do not need to be configured in two places (config and CSS).

WinnerWind commented:
> Small remark on the docs.
> 
> As a final thought: in the badges configuration, perhaps another, optional parameter could be added to specify/override the color of the badge of that type? That way, user-defined badges do not need to be configured in two places (config and CSS).

I'm unsure what you mean. User defined badges will auto generate CSS based on the `color` section. That is, if you define a new badge via `quartz.config.ts`, new CSS is generated specifically for that badge.

WinnerWind committed:
Fix tab spacing in documentation

Co-authored-by: Emile Bangma <github@emilebangma.com>

WinnerWind committed:
Define the type for BADGE_TYPES and allBadges

WinnerWind commented:
Also I'm unsure how to deal with this merge conflict. Sorry!

Edit: solved them, but it was quite messy.

WinnerWind committed:
Merge branch 'v4' into badges-plugin

WinnerWind committed:
Remove trailing comma from package.json

This prevented checks from passing.
